File: src/cloudtwin/providers/aws/bedrock/state.py
"""
Per-model request counter state for Bedrock simulation.

Thread-safe using a standard threading.Lock so the counter can be
incremented from concurrent async handlers without data races.
"""

from __future__ import annotations

import threading
from typing import Optional


class BedrockState:
    """Holds request counters keyed by modelId.

    ``increment`` returns the new (1-based) count so callers can use it
    immediately for sequence / error-injection decisions.
    """

    def __init__(self) -> None:
        self._counters: dict[str, int] = {}
        self._lock = threading.Lock()

    def increment(self, model_id: str) -> int:
        """Atomically increment the counter for *model_id* and return it."""
        with self._lock:
            count = self._counters.get(model_id, 0) + 1
            self._counters[model_id] = count
            return count

    def get_count(self, model_id: str) -> int:
        """Return the current counter for *model_id* without incrementing."""
        with self._lock:
            return self._counters.get(model_id, 0)

    def reset(self, model_id: Optional[str] = None) -> None:
        """Reset counters. Resets all models when *model_id* is None."""
        with self._lock:
            if model_id is None:
                self._counters.clear()
            else:
                self._counters.pop(model_id, None)

File: rust/cloudtwin-lite/src/aws/bedrock/state.rs
//! Per-model request counter with interior mutability.
//!
//! Kept in-memory only (Bedrock simulation is stateless w.r.t SQLite).

use std::collections::HashMap;
use std::sync::Mutex;

use super::config::BedrockSimConfig;

/// Thread-safe per-model request counters used by the scenario engine.
pub struct BedrockState {
    pub config: BedrockSimConfig,
    counters: Mutex<HashMap<String, u64>>,
}

impl BedrockState {
    pub fn new(config: BedrockSimConfig) -> Self {
        Self { config, counters: Mutex::new(HashMap::new()) }
    }

    /// Increment the counter for `model_id` and return the **new** count.
    pub fn increment(&self, model_id: &str) -> u64 {
        let mut lock = self.counters.lock().unwrap();
        let slot = lock.entry(model_id.to_string()).or_insert(0);
        *slot += 1;
        *slot
    }

    /// Return the current counter without modifying it.
    pub fn get_count(&self, model_id: &str) -> u64 {
        let lock = self.counters.lock().unwrap();
        lock.get(model_id).copied().unwrap_or(0)
    }

    /// Reset counters.  `None` resets all models; `Some(id)` resets one.
    pub fn reset(&self, model_id: Option<&str>) {
        let mut lock = self.counters.lock().unwrap();
        match model_id {
            None => lock.clear(),
            Some(id) => { lock.remove(id); }
        }
    }
}
